src/uct/ib/ud/base/ud_ep.c

```diff
@@ -1147,9 +1144,18 @@ uct_ud_ep_do_pending(ucs_arbiter_t *arbiter, ucs_arbiter_elem_t *elem,
         }
         return UCS_ARBITER_CB_RESULT_REMOVE_ELEM;
     }
+
     /* try to send ctl messages */
     uct_ud_ep_do_pending_ctl(ep, iface);
-    return uct_ud_ep_ctl_op_next(ep);
+    if (in_async_progress) {
+        return uct_ud_ep_ctl_op_next(ep);
+    } else {
```

Comment:
why do you need this `if else`? If there is any control operation `uct_ud_ep_ctl_op_next` would return UCS_ARBITER_CB_RESULT_NEXT_GROUP anyway.

Reply:
aa, the ctl ops might have been cleared during `uct_ud_ep_do_pending_ctl` handling, right?

Reply:
yes, will add comment